// vitest/dist/chunks/index.DdgEv5B1.js
import { v as vi, N as NodeBenchmarkRunner, S as Snapshots, T as TestRunner, a as assert, c as createExpect, g as globalExpect, i as inject, s as should, b as vitest } from './test.DNmyFkvJ.js';
import { b as bench } from './benchmark.CX_oY03V.js';
import { V as VitestEvaluatedModules } from './evaluatedModules.Dg1zASAC.js';
import { expectTypeOf } from 'expect-type';
import { afterAll, afterEach, aroundAll, aroundEach, beforeAll, beforeEach, describe, it, onTestFailed, onTestFinished, recordArtifact, suite, test } from '@vitest/runner';
import { chai } from '@vitest/expect';

const assertType = function assertType() {};

var index = /*#__PURE__*/Object.freeze({
  __proto__: null,
  BenchmarkRunner: NodeBenchmarkRunner,
  EvaluatedModules: VitestEvaluatedModules,
  Snapshots: Snapshots,
  TestRunner: TestRunner,
  afterAll: afterAll,
  afterEach: afterEach,
  aroundAll: aroundAll,
  aroundEach: aroundEach,
  assert: assert,
  assertType: assertType,
  beforeAll: beforeAll,
  beforeEach: beforeEach,
  bench: bench,
  chai: chai,
  createExpect: createExpect,
  describe: describe,
  expect: globalExpect,
  expectTypeOf: expectTypeOf,
  inject: inject,
  it: it,
  onTestFailed: onTestFailed,
  onTestFinished: onTestFinished,
  recordArtifact: recordArtifact,
  should: should,
  suite: suite,
  test: test,
  vi: vi,
  vitest: vitest
});

export { assertType as a, index as i };
